### PR TITLE
Index caching

### PR DESCRIPTION
**Pull request by kurransingh**

### CODE REVIEW

File: src/DCSAM.cpp [hunk 1]
```diff
@@ -137,32 +145,36 @@ void DCSAM::updateContinuous() {
...
-        boost::dynamic_pointer_cast<DCContinuousFactor>(graph[j]);
-    if (dcContinuousFactor) {
-      dcContinuousFactor->updateDiscrete(discreteVals);
-      for (const gtsam::Key &k : dcContinuousFactor->keys()) {
-        newAffectedKeys[j].insert(k);
-      }
+      boost::static_pointer_cast<DCContinuousFactor>(dcContinuousFactors_[j]);
+    dcContinuousFactor->updateDiscrete(discreteVals);
+    for (const gtsam::Key &k : dcContinuousFactor->keys()) {
+      // newAffectedKeys[dcIdxToFactor_.at(j)].insert(k);
```
Comment (keevindoherty) [inline src/DCSAM.cpp:155]: Can we remove this commented line if unneeded?

File: include/dcsam/DCSAM.h [hunk 2]
```diff
@@ -217,6 +217,7 @@ class DCSAM {
   // NOTE: will be needed for more sophisticated iSAM2 bookkeeping (rather than
   // naively iterating over all factors as we do now)
   std::vector<DCContinuousFactor::shared_ptr> dcContinuousFactors_;
+  gtsam::FastVector<gtsam::DiscreteFactor::shared_ptr> dcDiscreteFactors_;
   std::map<size_t, gtsam::FactorIndex> dcIdxToFactor_;
```
Comment (keevindoherty) [inline include/dcsam/DCSAM.h:221]: Maybe this can be removed if unneeded?

File: include/dcsam/DCSAM.h [hunk 3]
```diff
@@ -217,6 +217,7 @@ class DCSAM {
   // NOTE: will be needed for more sophisticated iSAM2 bookkeeping (rather than
   // naively iterating over all factors as we do now)
```
Comment (keevindoherty) [inline include/dcsam/DCSAM.h:218]: Can be removed.

File: src/DCSAM.cpp [hunk 4]
```diff
@@ -76,8 +78,10 @@ void DCSAM::update(const gtsam::NonlinearFactorGraph &graph,
     // NOTE: I think maybe this should be a boost::shared_ptr to avoid copy
     // construction
```
Comment (keevindoherty) [inline src/DCSAM.cpp:79]: Can be removed.

File: src/DCSAM.cpp [hunk 5]
```diff
@@ -115,13 +119,17 @@ void DCSAM::updateDiscreteInfo(const gtsam::Values &continuousVals,
...
     boost::shared_ptr<DCDiscreteFactor> dcDiscrete =
-        boost::dynamic_pointer_cast<DCDiscreteFactor>(dfg_[j]);
+        boost::static_pointer_cast<DCDiscreteFactor>(factor);
     if (dcDiscrete) {
       dcDiscrete->updateContinuous(continuousVals);
       dcDiscrete->updateDiscrete(discreteVals);
     }
+    else {
+      std::cout << "something is wrong with DCDiscreteFactors" << std::endl;
+    }
```
Comment (keevindoherty) [inline src/DCSAM.cpp:132]: `dcDiscreteFactors_` is a vector of pointers to `DCDiscreteFactor`s, so I don't think the explicit check if `dcDiscrete` is null is needed here (as it was for the previous `dynamic_pointer_cast`. Feel free to check me on this, but I think the static pointer cast will only result in a `nullptr` if `factor` itself is a `nullptr`.

File: src/DCSAM.cpp [hunk 6]
```diff
@@ -115,13 +119,17 @@ void DCSAM::updateDiscreteInfo(const gtsam::Values &continuousVals,
   if (continuousVals.empty()) return;
   // TODO(any): inefficient, consider storing indices of DCFactors
   // Update the DC factors with new continuous information.
```
Comment (keevindoherty) [inline src/DCSAM.cpp:121]: Can be removed.

File: src/DCSAM.cpp [hunk 7]
```diff
@@ -137,32 +145,36 @@ void DCSAM::updateContinuous() {
 void DCSAM::updateContinuousInfo(const DiscreteValues &discreteVals,
```
Comment (keevindoherty) [inline src/DCSAM.cpp:145]: Can remove preceding comment ("NOTE: for iSAM(2) requires special procedure [...]")

**Comment by keevindoherty** [comment]
Awesome, looking good and tests seem to be passing. Thanks for this @kurransingh . Comments are just related to some minor refactoring.

File: include/dcsam/DCSAM.h (hunk 2 above)
Comment (kurransingh) [reply include/dcsam/DCSAM.h:221]: Assuming you meant line 221 but yeah it is not needed, thanks!

**Pull request merged**